model/cluster_states.go

```diff
@@ -0,0 +1,173 @@
+}
+
+// AllCusterStatesPendingWork is a list of all cluster states that the supervisor
+// will attempt to transition towards stable on the next "tick".
+// Warning:
+// When creating a new cluster state, it must be added to this list if the cloud
+// cluster supervisor should perform some action on its next work cycle.
+var AllCusterStatesPendingWork = []string{
+	ClusterStateCreationRequested,
+	ClusterStateProvisioningRequested,
+	ClusterStateUpgradeRequested,
+	ClusterStateDeletionRequested,
+}
+
+// AllClusterRequestStates is a list of all states that a cluster can be put in
+// via the API.
+// Warning:
+// When creating a new cluster state, it must be added to this list if an API
+// endpoint should put the cluster in this state.
+var AllClusterRequestStates = []string{
+	ClusterStateCreationRequested,
+	ClusterStateProvisioningRequested,
+	ClusterStateUpgradeRequested,
+	ClusterStateDeletionRequested,
+}
+
+// ValidTransitionState returns whether a cluster can be transitioned into the
+// new state or not based on its current state.
+func (c *Cluster) ValidTransitionState(newState string) bool {
+	switch newState {
+	case ClusterStateCreationRequested:
+		return validTransitionToClusterStateCreationRequested(c.State)
+	case ClusterStateProvisioningRequested:
+		return validTransitionToClusterStateProvisioningRequested(c.State)
+	case ClusterStateUpgradeRequested:
+		return validTransitionToClusterStateUpgradeRequested(c.State)
+	case ClusterStateDeletionRequested:
+		return validTransitionToClusterStateDeletionRequested(c.State)
+	}
+
+	return false
+}
+
+func validTransitionToClusterStateCreationRequested(currentState string) bool {
+	switch currentState {
+	case ClusterStateCreationRequested,
+		ClusterStateCreationFailed:
+		return true
+	}
+
+	return false
+}
+
+func validTransitionToClusterStateProvisioningRequested(currentState string) bool {
+	switch currentState {
+	case ClusterStateStable,
+		ClusterStateProvisioningFailed,
+		ClusterStateProvisioningRequested:
+		return true
+	}
+
+	return false
+}
+
+func validTransitionToClusterStateUpgradeRequested(currentState string) bool {
+	switch currentState {
+	case ClusterStateStable,
+		ClusterStateUpgradeRequested,
+		ClusterStateUpgradeFailed:
+		return true
+	}
+
+	return false
+}
+
+func validTransitionToClusterStateDeletionRequested(currentState string) bool {
+	switch currentState {
+	case ClusterStateStable,
+		ClusterStateCreationRequested,
+		ClusterStateCreationFailed,
+		ClusterStateProvisioningFailed,
+		ClusterStateUpgradeRequested,
+		ClusterStateUpgradeFailed,
+		ClusterStateDeletionRequested,
+		ClusterStateDeletionFailed:
+		return true
+	}
+
+	return false
+}
+
+// ClusterStateReport is a report of all cluster requests states.
+type ClusterStateReport []StateReportEntry
+
+// StateReportEntry is a report entry of a given request state.
+type StateReportEntry struct {
+	RequestedState string
+	ValidStates    StateList
+	InvalidStates  StateList
```

Comment:
What about going with `CurrentState`, `ValidNextStates`, `InvalidNextStates`? I think that might make it clearer what's being represented here

Reply:
So, I do like this idea, but this could be a bit tricky in that the current state of an object could be something that the supervisor is currently working on and there could be many more states before the object is stable again. This pathway is something even the refactor doesn't capture and still is managed in the supervisor package.

Right now, the Report will tell you what states the object can/can't be in in regards to making a successful API request, which we already enforced. Perhaps I can think through the "supervisor pathway" question and do a follow-up PR sometime?

Thoughts?

Reply:
Ok to confirm, the report data is telling me if I were to request a state for an object, the `ValidStates` are the states the object must be in for me to be able to make that request. If the object is instead in a `InvalidStates` then my API request would fail.

An example:

I want to retry creation of a cluster through the API. I can only do that if my cluster is currently in the `creation-requested` or `creation-failed` states.

Does that sound right?

As for the supervisor pathway, I do think making that clearer in the future would be helpful but not entirely sure if moving that data/logic out of the supervisor makes sense. Maybe to expose it in the API like the state data here I guess

Reply:
Yep, that is correct!

And I agree on the supervisor stuff.